ENH: Add Markups ROI node

Sunderlandkyl opened the pull request:
Adds an ROI markup type with the vtkMRMLMarkupsROINode class.
ROI dimensions can be changed by clicking on the interactive handles on the faces and corners of the ROI.

The ROIType property of vtkMRMLMarkupsROINode controls how the ROI is defined:
- BOX: Users can place two control points to define the size of the ROI. Once the two points are placed, they are removed.
- BOUNDING_BOX: ROI encompasses all of its control points.

This commit also changes the ROI node type used in the Volume Rendering module from vtkMRMLAnnotationROINode to vtkMRMLMarkupsROINode.

Sunderlandkyl commented:
@lassoan This PR changes the volume rendering ROI from Annotation to Markups.
If we want to test the ROI on its own for a while, I can move that change to a separate PR.

lassoan commented:
Awesome, thank you, I'll review this today.

In the meantime, could you implement support for both the old and new ROI (so that both annotation and markups ROI node can be selected as ROI and used for cropping) and make it possible to change the volume rendering logic to create markups ROI depending on an application setting (or some Python code that can added to .slicerrc.py) to allow testing? Transition could then happen in phases:
- allow usage of both ROI classes, use annotation ROI by default (first integrated version)
- make volume rendering logic create markups ROI by default (after a couple of weeks)
- remove annotation module GUI, convert annotation ROI to markups ROI on scene import, update CLI module logic to use markups instead of annotation (in a few months)

Sunderlandkyl commented:
@lassoan Ok, vtkMRMLAnnotationROINode is still used by default, but both ROI types are supported.
Default ROI type can be changed with this line:

```slicer.modules.volumerendering.logic().SetDefaultROIClassName("vtkMRMLMarkupsROINode")```

lassoan commented on Modules/Loadable/Markups/Widgets/DesignerPlugins/qMRMLMarkupsROIWidgetPlugin.cxx:
```diff
@@ -0,0 +1,62 @@
+#include "qMRMLMarkupsROIWidget.h"
+
+//-----------------------------------------------------------------------------
+qMRMLMarkupsROIWidgetPlugin::qMRMLMarkupsROIWidgetPlugin(QObject *_parent)
```
Very nice, it is useful to have a reusable widget for this.

lassoan commented on Modules/Loadable/VolumeRendering/Logic/vtkSlicerVolumeRenderingLogic.cxx:
```diff
@@ -764,27 +768,48 @@ void vtkSlicerVolumeRenderingLogic::CopyLabelMapDisplayToVolumeRenderingDisplayN
+      xyz[i] *= 0.5;
+      }
+
+    markupsROINode->GetROIToLocalMatrix()->Identity();
```
Markups ROI should be rotated to match the volume axes.

Fitting ROI to a volume node could be a method in markups module logic (or maybe markups ROI node), as it can be useful for other modules, too.

lassoan reviewed (approved):
Wow, huge effort and very well done!

I did not find anything that would prevent integration today so that users can start testing it. If you don't have anything that you would like to change right now then go on and merge it.

I've added a couple of suggestions as comments and there are things that we should address in the coming weeks:

High priority:
- [x] Storage: use a single "orientation" matrix for storing orientation (instead of separate xAxis, yAxis, zAxis"), as it is more consistent with how we usually manage orientation
- [x] Storage: "sideLenghts" is very clear definition for a box, but not so good if we generalize the ROI (e.g., to ellipsoid); "diameter" or "size" could be better, as they are probably still clear enough and applicable to more shapes than just a box
- [x] Storage: "origin" is consistent with how we represent a coordinate system, but in case of a ROI, it may be more informative to have the "center" displayed in the file (if users create a file manually or from an external software, it could be more intuitive if they specify the center); "center" is also a bit more clear than "origin" (it may not be clear for everyone that origin is in a corner)
- [x] Storage: it would be better to display ROI properties (origin, size, etc.) near the beginning (before control points, measurments, display sections), because these files will be directly viewed and edited by users, so we should always start with the most important information
- [ ] Make filling 20% opacity by default (and set it to 0% when used as volume rendering clipping ROI)
Translation/rotation handles are too small in 3D (about half the size compared to slice views)
- [ ] When a ROI is already the active widget and user clicks place mode again then a new widget should be created (now it resets the current widget). At least until we improve the widget toolbar (we should have separate options for editing selected segmentation or create a new one).
- [ ] When the ROI is in bounding box mode and has many control points and then you switch to box mode then the ROI seems to remain in bounding box mode.
- [ ] We need outline (wireframe) display in 3D view (for example, for volume rendering clipping that would be a good representation)
- [ ] After placing a ROI (at least when it is done, interactively, initiated via the GUI) then make the interaction handles displayed by default, because now after you place the ROI it is very hard to adjust it.
- [ ] Make right-click menu work on all interaction handle points
- [ ] Make right-click work on the outline (but maybe on the filled area as well, if it is >0%) so that interaction handles can be activated/deactivated in a viewer (it can be hard to find a ROI node in the Data module)
- [ ] Add "Interaction handles visibility" option to eye icon right-click menu in subject hierarchy tree (probably for all markups)

Low priority:
- [ ] It is odd that there are no control points at allin box mode. Could we store at least the center as a control point?
- [ ] Interaction is very slow in debug mode. It seems to be due to plane clipping. It couldbe nicer if we could use one clipping filter because this filter has huge overhead when clipping very simple shapes. It could be also faster to use vtkPlaneCutter.
- [ ] When interacting with the ROI in one-up 3D view, the slice representation is still being updated (vtkSlicerROIRepresentation2D::UpdateFromMRML is called and it runs this->ROIWorldToSliceTransformFilter->Update(), which takes a long time)
- [ ] How does the ROI work with non-linear transforms? It should probably work similarly to angle and line markup, so that it would not be warped by the transforms, only the control points/handles would be transformed and they would be connected with straight lines.
- [ ] When moving the mouse over an interaction handle, the mouse cursor and handle color alternates between hand and arrow and selected/unselected color. It always ends up being correct but the flickering is distracting.
- [ ] It would be probably better if all interaction handles disappear when the ROI is out of the slice plane (now the arrows are always displayed)
- [ ] Add a shortcut (Ctrl+Left-click-and-drag) for symmetric adjustment of handle. This allows making the ROI box thinner without changing its position.
- [ ] Can you add an ellipsoid mode? It would help identifying any design issues that may come up in the future when we want to generalize the concept (e.g., create custom ROI widgets). Handles could be probably the same as now (no need to )
- [ ] If a bounding box type ROI is created and two points are placed then the ROI is not filled (even though it has 0 thickness, it should be filled)
- [ ] We could consider initializing the ROI with single-slice thickness. This would facilitate using it for generating 2D-like ROIs (useful for generating training data for object detection AI applications)
- [ ] Show node name in views (it could be similar how I enabled label display recently for curves, but in this case the label should be displayed in the center)
- [ ] Add measurement plugins (volume, diagonal, maybe xyz diameter)

lassoan commented:
Just one thing: I've got a compilation error for this line:

```
  vtkGetObjectMacro(ROIToLocalMatrix, vtkMatrix4x4);
```

As a quick workaround, I've replaced it by:

```
  vtkMatrix4x4* GetROIToLocalMatrix()
  {
    return this->ROIToLocalMatrix;
  }
```

Have you encountered this issue?

lassoan commented:
One more important thing: in vtkMRMLMarkupsROIJsonStorageNode::vtkInternalROI::UpdateMarkupsNodeFromJsonValue we need to convert the direction from RAS to LPS if the coordinate system is set to LPS (LPS is the default).

Sunderlandkyl commented:
> Just one thing: I've got a compilation error for this line:
> 
> ```
>   vtkGetObjectMacro(ROIToLocalMatrix, vtkMatrix4x4);
> ```
> 
> As a quick workaround, I've replaced it by:
> 
> ```
>   vtkMatrix4x4* GetROIToLocalMatrix()
>   {
>     return this->ROIToLocalMatrix;
>   }
> ```
> 
> Have you encountered this issue?

Interesting. No, I haven't encountered it. I'll see if I can reproduce it.

lassoan commented:
> > Just one thing: I've got a compilation error for this line:
> > ```
> >   vtkGetObjectMacro(ROIToLocalMatrix, vtkMatrix4x4);
> > ```
> > 
> > 
> > As a quick workaround, I've replaced it by:
> > ```
> >   vtkMatrix4x4* GetROIToLocalMatrix()
> >   {
> >     return this->ROIToLocalMatrix;
> >   }
> > ```
> > 
> > 
> > Have you encountered this issue?
> 
> Interesting. No, I haven't encountered it. I'll see if I can reproduce it.

It only happens in debug mode, very strange:

```
27>------ Build started: Project: vtkSlicerMarkupsModuleMRML, Configuration: Debug x64 ------
27>For vtkSlicerMarkupsModuleMRML - updating vtkSlicerMarkupsModuleMRMLHierarchy.txt
27>vtkMRMLMarkupsROINode.cxx
27>C:\D\S4\Modules\Loadable\Markups\MRML\vtkMRMLMarkupsROINode.h(105,3): error C2666: 'vtkOStreamWrapper::operator <<': 5 overloads have similar conversions
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(155,22): message : could be 'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkOStreamWrapper::std_string &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(107,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(void (__cdecl *)(void *,int *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(106,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const char *(__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(105,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(float *(__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(104,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(int *(__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(103,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(int (__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(102,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(void *(__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(101,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(void (__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(87,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(bool)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(86,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(double)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(85,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(float)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(84,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(unsigned __int64)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(83,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(unsigned long)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(82,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(unsigned int)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(81,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(unsigned short)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(80,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(unsigned char)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(79,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(__int64)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(78,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(long)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(77,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(int)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(76,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(short)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(75,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(char)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(74,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(void *)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(73,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const char *)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(72,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkStdString &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(71,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkSmartPointerBase &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(70,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkLargeInteger &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(69,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(vtkObjectBase &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(68,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkIndent &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(67,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkOStreamWrapper::EndlType &)'
27>C:\D\S4D\VTK\Common\Core\vtkSmartPointerBase.h(130,31): message : or       'std::ostream &operator <<(std::ostream &,const vtkSmartPointerBase &)'
27>C:\D\S4D\VTK\Common\Core\vtkVariant.h(63,31): message : or       'std::ostream &operator <<(std::ostream &,const vtkVariant &)'
27>C:\D\S4D\VTK\Common\Core\vtkStdString.h(32,31): message : or       'std::ostream &operator <<(std::ostream &,const vtkStdString &)'
27>C:\D\S4D\VTK\Common\Core\vtkWeakPointerBase.h(121,31): message : or       'std::ostream &operator <<(std::ostream &,const vtkWeakPointerBase &)'
27>C:\D\S4D\VTK\Common\Core\vtkObjectBase.h(216,40): message : or       'std::ostream &operator <<(std::ostream &,vtkObjectBase &)'
27>C:\D\S4D\VTK\Common\Core\vtkIndent.h(31,31): message : or       'std::ostream &operator <<(std::ostream &,const vtkIndent &)'
27>C:\D\S4D\VTK\Common\Core\vtkSmartPointer.h(248,17): message : or       'std::ostream &operator <<<vtkMatrix4x4>(std::ostream &,const vtkSmartPointer<vtkMatrix4x4> &)'
27>C:\D\S4\Modules\Loadable\Markups\MRML\vtkMRMLMarkupsROINode.h(105,3): message : while trying to match the argument list '(vtkOStreamWrapper, vtkSmartPointer<vtkMatrix4x4>)'
27>vtkMRMLMarkupsROIJsonStorageNode.cxx
27>C:\D\S4\Modules\Loadable\Markups\MRML\vtkMRMLMarkupsROINode.h(105,3): error C2666: 'vtkOStreamWrapper::operator <<': 5 overloads have similar conversions
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(155,22): message : could be 'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkOStreamWrapper::std_string &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(107,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(void (__cdecl *)(void *,int *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(106,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const char *(__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(105,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(float *(__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(104,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(int *(__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(103,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(int (__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(102,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(void *(__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(101,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(void (__cdecl *)(void *))'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(87,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(bool)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(86,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(double)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(85,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(float)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(84,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(unsigned __int64)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(83,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(unsigned long)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(82,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(unsigned int)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(81,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(unsigned short)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(80,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(unsigned char)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(79,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(__int64)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(78,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(long)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(77,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(int)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(76,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(short)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(75,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(char)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(74,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(void *)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(73,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const char *)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(72,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkStdString &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(71,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkSmartPointerBase &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(70,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkLargeInteger &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(69,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(vtkObjectBase &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(68,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkIndent &)'
27>C:\D\S4D\VTK\Common\Core\vtkOStreamWrapper.h(67,22): message : or       'vtkOStreamWrapper &vtkOStreamWrapper::operator <<(const vtkOStreamWrapper::EndlType &)'
27>C:\D\S4D\VTK\Common\Core\vtkSmartPointerBase.h(130,31): message : or       'std::ostream &operator <<(std::ostream &,const vtkSmartPointerBase &)'
27>C:\D\S4D\VTK\Common\Core\vtkVariant.h(63,31): message : or       'std::ostream &operator <<(std::ostream &,const vtkVariant &)'
27>C:\D\S4D\VTK\Common\Core\vtkStdString.h(32,31): message : or       'std::ostream &operator <<(std::ostream &,const vtkStdString &)'
27>C:\D\S4D\VTK\Common\Core\vtkWeakPointerBase.h(121,31): message : or       'std::ostream &operator <<(std::ostream &,const vtkWeakPointerBase &)'
27>C:\D\S4D\VTK\Common\Core\vtkObjectBase.h(216,40): message : or       'std::ostream &operator <<(std::ostream &,vtkObjectBase &)'
27>C:\D\S4D\VTK\Common\Core\vtkIndent.h(31,31): message : or       'std::ostream &operator <<(std::ostream &,const vtkIndent &)'
27>C:\D\S4D\VTK\Common\Core\vtkSmartPointer.h(248,17): message : or       'std::ostream &operator <<<vtkMatrix4x4>(std::ostream &,const vtkSmartPointer<vtkMatrix4x4> &)'
27>C:\D\S4\Modules\Loadable\Markups\MRML\vtkMRMLMarkupsROINode.h(105,3): message : while trying to match the argument list '(vtkOStreamWrapper, vtkSmartPointer<vtkMatrix4x4>)'
27>Generating Code...
27>Done building project "vtkSlicerMarkupsModuleMRML.vcxproj" -- FAILED.
```

Sunderlandkyl commented:
@lassoan I've addressed most of the comments that would break saving/loading compatibility down the line, as well the change to fix the compile error in Debug mode.

If it looks good then this can be integrated and I'll make the rest of the changes soon in a follow-up PR.

lassoan reviewed (approved):
Great, thanks a lot! If CircleCI passes then it is ready to be merged.

lassoan commented:
By the way, what was the issue with   `vtkGetObjectMacro(ROIToLocalMatrix, vtkMatrix4x4)`?

Sunderlandkyl commented:
I don't know, I just added the workaround that you mentioned.
I'll make a debug build tonight and see if I can find out what the issue is.
